File: src/algorithms/arrandstring/CombinationSum.java
package algorithms.arrandstring; 

import java.util.*; 

public class CombinationSum {
    /* Function: Given an array and a target, find all solutions to sum up to target. 
     * Solution: DFS. Notice when target < 0, can return. 
     */ 
    
    public List<List<Integer>> combinationSum2(int[] num, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>(); 
        if (num == null || num.length == 0) {
            return res; 
        }
        
        Arrays.sort(num); 
        List<Integer> list = new ArrayList<Integer>(); 
        sumHelper(num, target, 0, list, res); 
        return res; 
    }
    
    private void sumHelper(int[] num, int target, int st, List<Integer> list, List<List<Integer>> res) {
        if (target == 0) {
            res.add(new ArrayList<Integer>(list)); 
        } else if (target < 0) {
            return; 
        }
        
        for (int i = st; i < num.length; ++i) {
            if (i != st && num[i] == num[i - 1]) {
                continue; 
            }
            
            list.add(num[i]); 
            sumHelper(num, target - num[i], i + 1, list, res); 
            list.remove(list.size() - 1); 
        }
    }
}

File: src/CPP/src/leetcode/CombinationSum.cpp
/*
 * [Source] https://leetcode.com/problems/combination-sum/
 * [Difficulty]: Medium
 * [Tag]: Array
 * [Tag]: Backtracking
 */

#include <iostream>

using namespace std;

// [Solution]: Sort. Add current until larger than target, then remove the last added and try next one.
// [Corner Case]: though same value can use repeatly, if array has dup, it will create dup results
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/arrandstring/CombinationSumRepeatChoose.java
 */

/* Java solution
public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> result=new ArrayList<List<Integer>>();
        if(candidates.length<=0 || candidates[0]>target) return result;
        List<Integer> list=new ArrayList<Integer>();
        combination(candidates, target, 0, 0, list, result);
        return result;
    }

    public void combination(int[] num, int target, int sum, int idx, List<Integer> prev, List<List<Integer>> result){
        if(sum>target) return;
        if(sum==target){
            if(!result.contains(prev)){
                List<Integer> list=new ArrayList<Integer>(prev);
                result.add(list);
            }
            return;
        }
        for(int i=idx; i<num.length; i++){
            prev.add(num[i]);
            combination(num, target, sum+num[i], i, prev, result);
            prev.remove(prev.size()-1);
        }
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
